sample/weighted.go

```diff
@@ -0,0 +1,133 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this code is governed by a BSD-style
+// license that can be found in the LICENSE file
+
+package sample
+
+import (
+	"math/rand"
+
+	"github.com/gonum/floats"
+)
+
+// Weighted provides sampling without replacement from a collection of items with
+// non-uniform probability.
+type Weighted struct {
+	weights []float64
+	// heap is a weight heap.
+	//
+	// It keeps a heap-organised sum of remaining
+	// index weights that are available to be taken
+	// from.
+	//
+	// Each elements holds the sum of weights for
+	// the corresponding index, plus the sum of
+	// of its children's weights; the children
+	// of an element i can be found at positions
+	// 2*(i+1)-1 and 2*(i+1). The root of the
+	// weight heap is at element 0.
+	//
+	// See comments in container/heap for an
+	// explanation of the layout of a heap.
+	heap []float64
+	src  *rand.Rand
+}
+
+// NewWeighted returns a Weighted for the weights w. If src is nil, rand.Rand is
+// used as the random source.
+func NewWeighted(w []float64, src *rand.Rand) Weighted {
+	s := Weighted{
+		weights: make([]float64, len(w)),
+		heap:    make([]float64, len(w)),
+	}
+	s.ReweightAll(w)
+	return s
+}
+
+// Len returns the number of items held by the Weighted, including items
+// already taken.
+func (s Weighted) Len() int { return len(s.weights) }
+
+// Take returns an index from the Weighted with probability proportional
+// to the weight of the item. The weight of the item is then set to zero.
+// Take returns false if there are no items remaining.
+func (s Weighted) Take() (idx int, ok bool) {
```

Comment:
I agree with the simple comment (a la signficant variance), although it's also true when the absolute value of the weights are tiny.

It's not that AbsOrRel won't work for 1e4, it's that the actual threshold value of 1e-12 may not be correct, maybe it should be 1e-10 instead. 

I'm thinking about this through the lens of what I encoutered in optimize. I wanted to avoid a large cause of linesearch failure by just saying that small variations in the function value (1e-14 or so) aren't relevant. The problem is that for any reasonable threshold one wants to define, there is always a function whose floating point noise is higher, particularly for concurrent functions. The ramifications were too complicated in optimize, so we just gave up. 

Here is different because we are only doing subtraction, but that was the motivation for the initial comment. Why 1e-12? Why not 1e-6 or 1e-14? Because it's a reasonable default if the weights are reasonable, which means roughly O(1) and not O(1e-10) or O(1e10)